Improve variable terminal width support

Rohitth007 opened the pull request:
<!-- Please
 see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->
This PR improves terminal width readability and support for both small and wide terminals.
<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->
Partial fix for #1005
<!-- Add a link to a discussion on chat.zulip.org, if relevant -->
[topic link](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Zooming)

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [x] Manually
- [x] Existing tests (adapted, if necessary)
- [x] New tests added (for any new behavior)
- [x] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
<!-- For example:
- first commit doing some thing
- maybe multiple commits doing similar things
-->
**Commit 1**: ui: Make ZT go to into autohide mode if terminal is small.
**Commit 2**: ui: Fix center panel width and add padding for wide terminals.

**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
<!-- eg.
- Waiting on #<PR>
- Blocks #<PR>
-->
Related PR #388 

**Visual changes** <!-- if any; add/delete/fill-in with screenshot/diagram as appropriate -->
To be added (many cases)

Rohitth007 committed:
run/ui: Add autohide_fluid layout option for variable panel overlay.

This commit make the side panels in autohide mode expand after
crossing MAXNORMALWIDTH (160) such that it occupies 20% of the frame
this stops after MAX_TOTAL_WIDTH (200) columns is reached.

Tests added.

neiljp reviewed (commented):
@Rohitth007 The first commit looks good behavior-wise and the use of render here is effective :tada:
Code-wise you have some tidying/simplification included that could be a refactor commit first?  I'll want to confirm how the auto-autohide interacts with the explicit settings before merging, and that we have tests passing on a commit basis too.

Style wise I'm not convinced by the design after the second commit for wide terminals - did we decide upon that somewhere that I don't remember? (This isn't web app behavior either?)

Rohitth007 commented on zulipterminal/ui.py:
```diff
@@ -297,6 +291,20 @@ def keypress(self, size: urwid_Box, key: str) -> Optional[str]:
+            if maxcols < 102 and not self.controller.pseudo_autohide:
+                self.controller.pseudo_autohide = True
+                self.show_right_panel(visible=False)
+                self.body.focus_position = 0
```
The problem is, if the user is in the right column and we decrease width the right column disappears so the user would not know where the focus is hence I set it to left column.
Maybe if we hide the column that is not in use currently then this problem would be solved.

Rohitth007 committed:
refactor: ui: Extract autohide into View to make the calls shorter.

This commit:
* stores `self.controller.autohide` as `self.autohide` to make the
attribute call shorter.

Tests updated.

Rohitth007 commented on zulipterminal/ui.py:
```diff
@@ -297,6 +291,20 @@ def keypress(self, size: urwid_Box, key: str) -> Optional[str]:
+                self.controller.pseudo_autohide = True
+                self.show_right_panel(visible=False)
+                self.body.focus_position = 0
+            elif maxcols > 102 and self.controller.pseudo_autohide:
+                self.show_right_panel(visible=True)
+                self.show_left_panel(visible=True)
+                self.controller.pseudo_autohide = False
+
```
What would happen in the `else` statement that you are suggesting? The second part of the conditional is basically so that every render call should not change body.contents. Most of the time the render call is going to enter `else`. (Many time actually)

Rohitth007 replied:
Nothing happens at 102 but there may be different behavior while increasing or decreasing width. It better to make it fall into the autohide category

Ezio-Sarthak replied:
Not sure but can we define a constant variable for 102? The variable name could justify many things

Rohitth007 replied:
Sure, I'll add them.

neiljp commented on zulipterminal/ui.py:
```diff
@@ -315,6 +316,25 @@ def mouse_event(
+        if not self.autohide:
+            if maxcols <= 102 and self.mode != "small":
+                self.mode = "small"
+                if self.focus_col != 2:
```
Are these checks why you're setting self.focus_col? They have no other functional purpose?

Rohitth007 replied:
They replace every `self.body.focus_position` there are about 17+ of those.

neiljp reviewed (commented):
@Rohitth007 You've not changed the label here, but you mentioned on the stream so I've done another review here.

It's good to see the not directly related changes in the first commit, but you've added others than I don't quite follow.

The behavior arising from the last commit looks better now, though we can discuss further in the stream whether we perhaps want a vertical line border close on either side (the edge of the users list looks somewhat odd otherwise?), and regarding the interaction between wide and autohide - possibly we could simply add a new 'dynamic' autohide value as the default, which would make this cleaner and retain backwards compatibility?

> Maybe if we hide the column that is not in use currently then this problem would be solved.

Is this supposed to be resolved?

> What would happen in the else statement that you are suggesting? The second part of the conditional is basically so that every render call should not change body.contents. Most of the time the render call is going to enter else. (Many time actually)

The else part would be a "this shouldn't happen" part :)
For the last commit it may look cleaner with a `if x < size_n` chain with increasing size_n?

Rohitth007 commented on zulipterminal/ui.py:
```diff
@@ -135,17 +136,18 @@ def main_window(self) -> Any:
         self.right_panel = self.right_column_view()
         if self.controller.autohide:
```
Oh I thought I had updated the rest of them as well in this commit, I might have missed them.

Rohitth007 commented on zulipterminal/ui.py:
```diff
@@ -135,17 +136,18 @@ def main_window(self) -> Any:
+                (self.RIGHT_WIDTH, self.right_panel),
             ]
         self.body = urwid.Columns(body, focus_column=0)
+        self.focus_col = self.body.focus_position
```
This is basically to make the setting of focus_position shorter so that the code is easier to read as it is used many times.
Similarly for autohide as well.

neiljp replied:
I understand it's useful for tracking the column, but it doesn't set the actually property. focus_col takes the value of body.focus_position, but setting the former again doesn't update the latter.

Rohitth007 replied:
But both of them point to the same property object right? So it would affect the object and they are not updating independent?

Rohitth007 replied:
I see it wasn't how I expected it to be. Is there a nice way to alias body.focus_position?

neiljp replied:
Inside the object you could use self.body.focus_position; if setting from outside too then it may be clearer to have a method for it?

Rohitth007 commented on zulipterminal/ui.py:
```diff
@@ -315,6 +316,25 @@ def mouse_event(
+                self.show_right_panel(visible=True)
+                self.show_left_panel(visible=True)
+                self.focus_col = prev_focus
+                self.mode = "normal"
```
`ignore_check` option wasn't made yet. So I had to do that. I have moved it to this commit now, so all of them will look similar

Rohitth007 commented:
I think having assert statements in else is unnecessary and disturbs the way it reads. Do we need it?
![image](https://user-images.githubusercontent.com/64144419/126459490-ef5d72c2-02e5-47bd-87ef-0c8ab37393c2.png)

Rohitth007 commented:
Re preserving focus, yes it is resolved.
Re vertical border, I'm don't think it can be changed dynamically, it has to be set once but that means horizontal space would be lost. Anyway, if we color the side panels differently (division based on color) then it should not look odd?

neiljp commented on zulipterminal/ui.py:
```diff
@@ -176,16 +181,16 @@ def main_window(self) -> Any:
 
-    def show_left_panel(self, *, visible: bool) -> None:
-        if not self.autohide:
+    def show_left_panel(self, *, visible: bool, ignore_check: bool = False) -> None:
+        if not self.autohide and self.mode != "small" and not ignore_check:
```
Can we use `visible` instead of `ignore_check`?

Rohitth007 replied:
We might want to hide (visible=False) but skip the check. That would not be possible right?

neiljp commented on zulipterminal/ui.py:
```diff
@@ -314,6 +319,25 @@ def mouse_event(
+        if not self.autohide:
+            if maxcols < MINNORMALWIDTH and self.mode != "small":
+                self.mode = "small"
+                if self.body.focus_position == 2:
+                    self.show_left_panel(visible=False)
+                else:
+                    self.show_right_panel(visible=False)
```
This works well, but perhaps if we're in the central panel we should close both sides?

Rohitth007 replied:
I think only hiding 1 panel is better? Or having both disappear might be confusing?

neiljp replied:
This will perhaps depend on the tab feature.

Rohitth007 replied:
Yes, and also not hiding both would confuse the user as the left column stays but the right column appears and disappears on right then left keypress. I have made both disappear.

neiljp commented on zulipterminal/ui.py:
```diff
@@ -133,17 +134,17 @@ def main_window(self) -> Any:
+        if self.autohide:
             body = [
-                (View.LEFT_WIDTH, self.left_panel),
+                (self.LEFT_WIDTH, self.left_panel),
```
There's a hybrid in the rest of the PR of View constants and module constants - let's keep to one or the other, perhaps a standard of module constants, starting in this commit?

Rohitth007 replied:
where else is `View` used? In tests View.LEFT_WIDTH has to be used because it is in parametrize.

neiljp reviewed (commented):
@Rohitth007 This works well, particularly the autohide which I almost merged - but we don't yet have tests (adapted, or new) and there are a few locations we could tidy.

I mentioned my logic re the asserts/exceptions in the stream, but basically it's a "code shouldn't reach here" check.

Rohitth007 commented on zulipterminal/ui.py:
```diff
@@ -21,6 +22,9 @@
 from zulipterminal.urwid_types import urwid_Box
 
 
+MINNORMALWIDTH = 103
```
The variable becomes too long and has to split into multiple lines hence I went with this

Rohitth007 committed:
run/ui: Add `dynamic` layout option for variable terminal support.

A dynamic option is added it ZT which adapts to various terminal
widths to make it more readable. To shift from using a bool to
options `self.autohide` to changed to `self.layout`.

This makes ZT behave like autohide when in a small terminal which is
defined using MINNORMALWIDTH. After crossing this limit ZT behaves
like no-autohide until MAXNORMALWIDTH after which, side panel widths
are not fixed anymore and they start occupying space in the ratio
20:60:20. This gives the effect that the side panels are expanding.
This stops when we reach the max width set for ZT.

To do so this makes use of urwid's render method to get maxcols, a
Literal `self.mode` which takes values "small", "normal" and "wide"
and existing `show_left_panel` and `show_right_panel` methods with
small changes.

Rohitth007 committed:
config: run/ui/view: Change `autohide` attribute to `layout`.

This commit changes the internal code attributes form
`autohide` to `layout`

Tests updated.

Rohitth007 committed:
ui: Add Padding and LineBox to centered Frame for wide terminals.

This commit adds Padding around a fixed total width of
MAX_TOTAL_WIDTH (200) of all panels combined. The padding is added
around the Frame instead of the body so that even the title and
footer have a fixed max width.

A LineBox is added around the padded_frame to give it borders in a
wide terminal.

self.has_border is used to make sure every render call does not
create a LineBox.

All this is done inside `add_padding_and_border_to_frame` method

Tests added.

Rohitth007 committed:
documentation: README: Add *layout* config to zuliprc doc.

This commit:
* Adds recently introduced *layout* config which defaults
to `dynamic` to the zuliprc template in README.
* Keeps backwards compatible *autohide* config.
* Adds spacing after each config.
* Comments out *footlinks* as both this and *maximum-footlinks*
can't be used together.

Rohitth007 committed:
config: run: Move from boolean autohide to layout options.

This commit moves from a self.autohide which was a bool to
self.layout which can take multiple options. This enables addition
of new panel layouts in the future.

Tests updated.

Rohitth007 committed:
refactor: ui/views/buttons: Combine autohide handling with focus.

This commit combines the focus setting with autohide handling. This
makes it easier to handle focus change of panels since both these
functions are often seen used together.

This also avoids the situation of the show_panels handling focus as
moving to a different panel through focus_panel takes care of that.

This also gives the advantage of knowing which panel we are on even
if we are in autohide layout hence giving us the ability to
manipulate the current panel (Eg: resizing) depending where we are.

Tests added.
